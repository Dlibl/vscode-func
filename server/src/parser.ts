import * as Parser from 'tree-sitter';
import * as FuncLanguage from 'tree-sitter-func-language';


export const createParser = () => {
    const parser = new Parser();
    parser.setLanguage(FuncLanguage);
    return parser
}