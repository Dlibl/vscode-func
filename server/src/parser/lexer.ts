import { createToken, Lexer } from 'chevrotain'


export const True = createToken({ name: "True", pattern: /true/ })
export const False = createToken({ name: "False", pattern: /false/ })
export const Null = createToken({ name: "Null", pattern: /null/ }) 
export const LCurly = createToken({ name: "LCurly", pattern: /{/ })
export const RCurly = createToken({ name: "RCurly", pattern: /}/ })
export const LSquare = createToken({ name: "LSquare", pattern: /\[/ })
export const RSquare = createToken({ name: "RSquare", pattern: /]/ })
export const LParen = createToken({ name: 'LParen', pattern: /\(/ })
export const RParen = createToken({ name: 'RParen', pattern: /\)/ })
export const Comma = createToken({ name: "Comma", pattern: /,/ })
export const Semicolon = createToken({ name: "Semicolon", pattern: /;/ })
export const GlobalModifier = createToken({ name: "GlobalModifier", pattern: /global/ })
export const AtomicType = createToken({ name: 'AtomicType', pattern: /var|int|slice|tuple|cell|builder|cont|_/ })
export const IfKeyword = createToken({ name: 'IfKeyword', pattern: /if/ })
export const AsmKeyword = createToken({ name: 'AsmKeyword', pattern: /asm/ })
export const ForallKeyword = createToken({ name: 'AsmKeyword', pattern: /forall/ })
export const FunctionModifier = createToken({ name: "FunctionModifier", pattern: /impure|method_id|inline|inline_ref/ })
export const FunctionArrowOperator = createToken({ name: 'FunctionArrowOperator', pattern: /\-\>/ })
export const EqualityOperator = createToken({ name: 'EqualityOperator', pattern: /=/ })
export const Operator = createToken({ name: 'Operator', pattern: /(\<==\>|([~\-=])|((\<\<|\>\>|\^\>\>|\^\<\<|([\*\/\%\&\+\|\^\<\>])|([~\^]\/)|([~\^\/]\%))=)|(\<\<|\>\>|\^\>\>|\^\<\<|([\*\/\%\&\+\|\^\<\>])|([~\^]\/)|([~\^\/]\%)))/ })
export const StringLiteral = createToken({
    name: "StringLiteral",
    pattern: /"(:?[^\\"]|\\(:?[bfnrtv"\\/]|u[0-9a-fA-F]{4}))*"/,

  })
export const NumberLiteral = createToken({
    name: "NumberLiteral",
    pattern: /-?(0|[1-9]\d*)(\.\d+)?([eE][+-]?\d+)?/
})
export const WhiteSpace = createToken({
    name: "WhiteSpace",
    pattern: /[ \t\n\r]+/,
    group: Lexer.SKIPPED
})
export const Identifier = createToken({
    name: "Identifier",
    pattern: /(`([^`]+)`|(?=_[\w\d]*)(_([^;,\[\]\(\)\s~.\{\}]+))|(?![_`][\w\d]*)([^;,\[\]\(\)\s~.\{\}]+))/
})

LCurly.LABEL = '{'
RCurly.LABEL = '}'
LSquare.LABEL = '['
RSquare.LABEL = ']'
LParen.LABEL = '('
RParen.LABEL = ')'
Comma.LABEL = ','
Semicolon.LABEL = ';'
FunctionArrowOperator.LABEL = '->'
EqualityOperator.LABEL = '='
IfKeyword.LABEL = 'if'


export const funcTokens = [
    True, False, Null, 
    LCurly, RCurly, LSquare,
    RSquare, LParen, RParen,
    Comma, Semicolon, StringLiteral,
    GlobalModifier, 
    NumberLiteral, WhiteSpace,
    AtomicType,
    EqualityOperator,
    FunctionArrowOperator,
    IfKeyword,
    AsmKeyword,
    ForallKeyword,
    FunctionModifier,
    Operator,
    Identifier,
]

export const FuncLexer = new Lexer(funcTokens, {
    positionTracking: "onlyOffset"
})
