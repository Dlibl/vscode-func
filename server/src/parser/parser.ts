import { CstParser } from 'chevrotain';
import { 
    AsmKeyword, AtomicType, Comma,
    EqualityOperator,
    FuncLexer, FunctionArrowOperator, 
    FunctionModifier, funcTokens, GlobalModifier, 
    Identifier, IfKeyword, LCurly, LParen, LSquare, 
    RCurly, RParen, RSquare, Semicolon, 
    StringLiteral
} from './lexer';


export class FuncParser extends CstParser {
    constructor() {
        super(funcTokens)
        this.performSelfAnalysis()
    }

    public func = this.RULE("func", () => {
        this.MANY({
            DEF: () => {
                this.OR([
                    { ALT: () => this.SUBRULE(this.function) },
                    { ALT: () => this.SUBRULE(this.globalVariable) }
                ])
            }
        })
    })

    //#region Root rules
    private function = this.RULE('function', () => {
        this.SUBRULE(this.typeDeclaration);
        this.CONSUME(Identifier);
        this.CONSUME(LParen);
        this.SUBRULE(this.functionParameters);
        this.CONSUME(RParen);
        this.MANY({
            DEF: () => this.CONSUME(FunctionModifier)
        })
        this.OR([
            {
                ALT: () => {
                    this.SUBRULE(this.blockStatement)
                }
            },
            {
                ALT: () => this.SUBRULE(this.asmReference)
            }
        ])
    })

    private globalVariable = this.RULE('globalVariable', () => {
        this.CONSUME(GlobalModifier);
        this.OR([
            {
                ALT: () => {
                    this.SUBRULE(this.typeDeclaration);
                    this.CONSUME(Identifier);
                }
            }, {
                ALT: () => {
                    this.CONSUME1(Identifier);
                }
            }
        ])
        this.CONSUME(Semicolon);
    })
    //#endregion

    private functionParameters = this.RULE('functionParameters', () => {
        this.MANY_SEP({
            SEP: Comma,
            DEF: () => {
                this.SUBRULE(this.parameterDeclaration);
            }
        })
    })

    private parameterDeclaration = this.RULE('parameterDeclaration', () => {
        this.SUBRULE(this.typeDeclaration);
        this.CONSUME(Identifier);
    })

    private asmReference = this.RULE('asmReference', () => {
        this.CONSUME(AsmKeyword);
        this.CONSUME(StringLiteral);
        this.CONSUME(Semicolon);
    })

    private blockStatement = this.RULE('blockStatement', () => {
        this.CONSUME(LCurly);
        this.MANY({
            DEF: () => this.SUBRULE(this.blockExpression)
        })
        this.CONSUME(RCurly);
    })
    private blockExpression = this.RULE('blockExpression', () => {
        this.OR([
            {
                ALT: () => {
                    this.SUBRULE(this.simpleExpression);
                    this.CONSUME(Semicolon);
                }             
            }, {
                ALT: () => {
                    this.SUBRULE(this.controlExpression);
                }
            }
        ])
    })


    //#region Simple expression
    private simpleExpression = this.RULE('simpleExpression', () => {
        this.OR([
            {
                ALT: () => this.SUBRULE(this.assignExpression)
            }, {
                ALT: () => this.SUBRULE(this.callExpression)
            }, {
                ALT: () => this.CONSUME(Identifier)
            }
        ])
    })

    private callExpression = this.RULE('callExpression', () => {
        this.CONSUME(Identifier);
        this.CONSUME(LParen);
        this.MANY_SEP({
            DEF: () => this.SUBRULE(this.simpleExpression),
            SEP: Comma,
        })
        this.CONSUME(RParen);
    })

    private assignExpression = this.RULE('assignExpression', () => {
        this.SUBRULE(this.identifierGroup);
        this.CONSUME(EqualityOperator);
        this.SUBRULE(this.simpleExpression);
    })

    private identifierGroup = this.RULE('identifierGroup', () => {
        this.OPTION({
            DEF: () => this.SUBRULE(this.typeDeclaration),
        })
        this.OR([
            {
                ALT: () => this.CONSUME(Identifier)
            },
            {
                ALT: () => {
                    this.CONSUME(LParen);
                    this.MANY_SEP({
                        SEP: Comma,
                        DEF: () => this.SUBRULE1(this.identifierGroup)
                    })
                    this.CONSUME(RParen);
                }
            },
            {
                ALT: () => {
                    this.CONSUME(LSquare);
                    this.MANY_SEP1({
                        SEP: Comma,
                        DEF: () => this.SUBRULE2(this.identifierGroup)
                    })
                    this.CONSUME(RSquare);
                }
            }
        ])
    })
    //#endregion

    //#region Control expression
    private controlExpression = this.RULE('controlExpression', () => {
        this.CONSUME(IfKeyword);
        this.SUBRULE(this.blockStatement);
    })
    //#endregion

    //#region Type declaration
    private typeDeclaration = this.RULE('typeDeclaration', () => {
        this.OR({
            DEF: [
                { ALT: () => this.CONSUME(AtomicType) },
                { ALT: () => this.SUBRULE(this.tupleType) },
                { 
                    ALT: () => this.SUBRULE(this.tensorType),
                    GATE: this.BACKTRACK(this.tensorType),
                },
                { 
                    ALT: () => this.SUBRULE(this.wrappedFunctionalType),
                    GATE: this.BACKTRACK(this.wrappedFunctionalType)
                }
            ]
        })
    })

    private tensorType = this.RULE('tensorType', () => {
        this.CONSUME(LParen);
        this.MANY_SEP({
            SEP: Comma,
            DEF: () => this.SUBRULE(this.typeDeclaration)
        });
        this.CONSUME(RParen);
    })

    private tupleType = this.RULE('tupleType', () => {
        this.CONSUME(LSquare)
        this.MANY_SEP({
            SEP: Comma,
            DEF: () => this.SUBRULE(this.typeDeclaration)
        })
        this.CONSUME(RSquare)
    })

    private wrappedFunctionalType = this.RULE('wrappedFunctionalType', () => {
        this.CONSUME(LParen);
        this.SUBRULE(this.functionalType);
        this.CONSUME(RParen);
    })

    private functionalType = this.RULE('functionalType', () => {
        this.SUBRULE(this.typeDeclaration);
        this.CONSUME(FunctionArrowOperator);
        this.SUBRULE1(this.typeDeclaration);
    })
    //#endregion
}


export function parseFunc(text: string) {
    let parser = new FuncParser()

    const lexResult = FuncLexer.tokenize(text)
    parser.input = lexResult.tokens
    const cst = parser.func()
  
    // this would be a TypeScript compilation error because our parser now has a clear API.
    // let value = parser.json_OopsTypo()
  
    return {
      cst: cst,
      lexErrors: lexResult.errors,
      parseErrors: parser.errors
    }
  }