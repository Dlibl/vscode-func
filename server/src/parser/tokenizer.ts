type Position = {
    index: number,
    line: number,
    character: number,
}

const TokenDictionary = {
    '(': 'lparen',
    ')': 'rparen',
    '{': 'lbrace',
    '}': 'rbrace',
    '[': 'lsqbracket',
    ']': 'rsqbracket',
    ',': 'comma',
    ';': 'semicolon',
} as const

export type Token = {
    position: Position,
    value: string,
    type: 'identifier' | (typeof TokenDictionary[keyof typeof TokenDictionary])
}

const Tokens = Object.keys(TokenDictionary)

class CharStream {
    private index = -1
    private currentLine = 0
    private currentCharacter = -1

    constructor(private code: string) {}

    public next() {
        let char = this.code[++this.index]
        if (char === '\n') {
            this.currentLine++
            this.currentCharacter = -1
        } else {
            this.currentCharacter++
        }
        return char
    }


    public position(): Position {
        return {
            index: this.index,
            line: this.currentLine,
            character: this.currentCharacter
        }
    }
}

export function tokenize(code: string): Token[] {
    let tokens: Token[] = []
    let iterator = new CharStream(code)
    let char: string | undefined = undefined

    let buffer = ''
    let bufferPosition: Position | undefined = undefined
    const flushBuffer = () => {
        if (buffer.length == 0) {
            return
        }

        tokens.push({
            position: bufferPosition,
            type: 'identifier',
            value: buffer
        })
        buffer = ''
        bufferPosition = undefined
    }
    while (char = iterator.next()) {
        if (char == ' ' || char == '\n' || Tokens.includes(char)) {
            flushBuffer()

            if (Tokens.includes(char)) {
                tokens.push({
                    position: iterator.position(),
                    type: TokenDictionary[char],
                    value: char
                })
            }
            continue
        }
        if (bufferPosition === undefined) {
            bufferPosition = iterator.position()
        }
        buffer += char 
    }
    flushBuffer()

    return tokens
}