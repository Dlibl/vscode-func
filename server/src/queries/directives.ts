import { parserQuery } from './utils/parserQuery';
import * as Parser from 'web-tree-sitter';

const query = parserQuery`
(include_directive) @include
(pragma_directive) @pragma
`

export function queryDirectives(node: Parser.SyntaxNode) {
    return query().captures(node);
}