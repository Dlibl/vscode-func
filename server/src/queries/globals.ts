import { parserQuery } from './utils/parserQuery';
import * as Parser from 'tree-sitter';

const query = parserQuery`
(function_definition name: (function_name) @function)
(global_var_declarations _ name: (identifier) @globalVar)
`

export function queryGlobals(node: Parser.SyntaxNode) {
    return query.captures(node).map(a => ({ text: a.node.text, type: a.name }));
}