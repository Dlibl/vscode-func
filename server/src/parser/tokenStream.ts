import { Token } from './tokenizer';

export class TokenStream {
    private pos = -1;

    constructor(private readonly tokens: Token[]) {}

    hasNext() {
        return this.pos+1 < this.tokens.length;
    }
    peek(): Token {
        return this.tokens[this.pos+1];
    }

    current(): Token {
        return this.tokens[Math.max(this.pos, 0)];
    }

    next(): Token {
        return this.tokens[++this.pos];
    }

    readUntil(...seps: Token['type'][]) {
        let tokens: Token[] = []
        let token: Token | undefined
        while (token = this.peek()) {
            if (seps.includes(token.type)) {
                break
            }
            tokens.push(token)
            this.next()
        }
        return tokens
    }
}