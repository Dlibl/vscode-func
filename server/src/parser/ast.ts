import { Token } from './tokenizer';
import { TokenStream } from './tokenStream';

type AstNode = Root | Identifier | Literal | VariableDeclaration | VariableDeclarator | FunctionDeclaration | BlockStatement | Modifier
interface Root { type: 'root', body: AstNode[] }
interface Identifier { type: 'identifier', name: string }
interface Modifier { type: 'modifier', name: string }
interface Literal { type: 'literal', name: string }

interface VariableDeclaration { 
    type: 'variable_declaration',
    declarations: VariableDeclarator[]
 }
interface VariableDeclarator {
    type: 'variable_declarator',
    id: Identifier,
    typeDef: Type
}
interface FunctionDeclaration { 
    type: 'function_declaration',
    id: Identifier,
    arguments: FunctionArguments,
    returnType: Type,
    modifiers: Modifier[]
    body: BlockStatement
}
interface FunctionArguments {
    type: 'function_arguments',
    declarations: VariableDeclarator[]
}
type Type = AtomicType | TensorType | TupleType | FunctionType | HoleType
interface TensorType {
    type: 'tensor_type'
    values: Type[]
}
interface TupleType {
    type: 'tuple_type'
    values: Type[]
}
interface FunctionType {
    type: 'function_type'
    args: Type
    returnType: Type
}
interface AtomicType {
    type: 'atomic_type'
    name: string
}
interface HoleType {
    type: 'hole_type'
}
interface BlockStatement {
    type: 'block_statement'
    body: AstNode[]
}

function parseIdentifier(tokens: TokenStream): Identifier {
    let id = tokens.next()
    if (id.type !== 'identifier') {
        // error
    }

    return {
        name: id.value,
        type: 'identifier'
    }
}

function parseArguments(tokens: TokenStream) {
    let args: FunctionArguments = {
        type: 'function_arguments',
        declarations: []
    }
    while (true) {
        let type = parseTypeDefinition(tokens)
        let id = parseIdentifier(tokens)
        args.declarations.push({
            id,
            typeDef: type,
            type: 'variable_declarator'
        })
        if (tokens.peek().type !== 'comma') {
            break
        }
        tokens.next()
    }
    return args
}

function parseVariableDeclarator(tokens: TokenStream): VariableDeclarator {
    let type = parseTypeDefinition(tokens);
    let identifier = parseIdentifier(tokens);

    return {
        id: identifier,
        typeDef: type,
        type: 'variable_declarator',
    }
}

function parseBlockStatement(tokens: TokenStream): BlockStatement {
    let bracketStack: ('{' | '(' | '[')[] = []
    if (tokens.peek().type !== 'lbrace') {
        console.log('invalid block statement')
    } else {
        tokens.next()
        bracketStack.push('{')
    }
    while (bracketStack.length !== 0) {
        let next = tokens.next()
        if (next.type === 'lbrace') {
            bracketStack.unshift('{')
        }
        if (next.type === 'lparen') {
            bracketStack.unshift('(')
        }
        if (next.type === 'lsqbracket') {
            bracketStack.unshift('[')
        }

        if (
            (bracketStack[0] === '{' && next.type === 'rbrace') ||
            (bracketStack[0] === '(' && next.type === 'rparen') ||
            (bracketStack[0] === '[' && next.type === 'rsqbracket')
        ) {
            bracketStack.shift()
        }
    }
    return {
        type: 'block_statement',
        body: []
    }
}

function parseTypeDefinition(tokens: TokenStream): Type {
    const atomicTypes = ['int', 'slice', 'cell', 'builder', 'tuple', 'cont', '_', 'var']
    let peeked = tokens.peek()
    if (peeked.type === 'identifier' && !atomicTypes.includes(peeked.value)) {
        return {
            type: 'hole_type'
        }
    }
    let token = tokens.next();
    if (token.type == 'identifier') {
        return {
            type: 'atomic_type',
            name: token.value
        }
    }
    if (token.type == 'lparen') {
        let next = tokens.peek()

        if (next.type == 'rparen') {
            // empty tensor type
            tokens.next()
            return {
                type: 'tensor_type',
                values: []
            }
        }

        let firstType = parseTypeDefinition(tokens)
        next = tokens.next()
        if (next.type === 'comma') { // tensor type
            let type: Type = {
                type: 'tensor_type',
                values: [firstType]
            }
            while (next.type !== 'rparen') {
                let nextType = parseTypeDefinition(tokens)
                type.values.push(nextType)
                next = tokens.next()
            }
            return type;
        } else if (next.type === 'rparen') {
            return {
                type: 'tensor_type',
                values: [firstType]
            }
        } else if (next.value === '->') { // function type
            let returnType = parseTypeDefinition(tokens)
            next = tokens.next()
            if (next.type !== 'rparen') {
                throw new Error('missing )')
            }

            return {
                type: 'function_type',
                args: firstType,
                returnType
            }
        }
    }
    if (token.type === 'lsqbracket') {
        let next = tokens.peek()
        if (next.type === 'rsqbracket') {
            tokens.next()

            return {
                type: 'tuple_type',
                values: []
            }
        }
        let type: Type = {
            type: 'tuple_type',
            values: []
        } 
        while (next.type !== 'rsqbracket') {
            let variable = parseTypeDefinition(tokens)
            type.values.push(variable)
            next = tokens.next()
            if (next.type !== 'comma' && next.type !== 'rsqbracket') {
                // error
            }
        }
        return type
    }
    console.log(`(${token.position.line}:${token.position.character}): cannot parse type definition: unexpected token: ${token.value}`)
}

function expectSemicolon(tokens: TokenStream) {
    let next = tokens.peek()
    if (next.type !== 'semicolon') {
        console.log(`(${next.position.line}:${next.position.character}): semicolon expected, got "${next.value}"`)
        return
    }
    tokens.next()
}

export function buildAstFromTokens(tokens: Token[]): AstNode {
    let root: Root = { type: 'root', body: [] }
    let path: AstNode[] = [root]
    let tokenStream = new TokenStream(tokens)
    while (tokenStream.hasNext()) {
        let current = path[0]

        if (current.type === 'root') {
            let next = tokenStream.peek()
            if (next.value === 'global') {
                tokenStream.next()

                // variable
                let nextNode: AstNode = {
                    type: 'variable_declaration',
                    declarations: [parseVariableDeclarator(tokenStream)]
                }
                expectSemicolon(tokenStream)

                current.body.push(nextNode)
            } else {
                let returnType = parseTypeDefinition(tokenStream)
                let name = tokenStream.next()
                if (name.type !== 'identifier') {
                    throw new Error('invalid function signature')
                }
                if (tokenStream.peek().type !== 'lparen') {
                    console.log('expected \'(\' in function declaration, not ' + tokenStream.peek().value);
                } else {
                    tokenStream.next()
                }

                let args = parseArguments(tokenStream)
                if (tokenStream.peek().type !== 'rparen') {
                    console.log('expected \')\' or \',\', not ' + tokenStream.peek().value);
                } else {
                    tokenStream.next()
                }
                let modifiers = tokenStream.readUntil('lbrace')

                let nextNode: AstNode = {
                    type: 'function_declaration',
                    arguments: args,
                    returnType,
                    id: {
                        name: name.value,
                        type: 'identifier'
                    },
                    modifiers: modifiers.map(a => ({
                        type: 'modifier',
                        name: a.value
                    })),
                    body: parseBlockStatement(tokenStream),
                }
                current.body.push(nextNode)
            }
        }
    }
    console.log(JSON.stringify(root))
    return root
}