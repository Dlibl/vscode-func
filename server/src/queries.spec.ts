import { createParser } from './parser'
import { queryGlobals } from './queries/globals';

it('should query globals', () => {
    let parser = createParser();
    let tree = parser.parse(`
        global var id;

        () test() {
            if (id == 1) {
                int i = 0;
            }
            int a = 0;
        }
    `)

    let globals = queryGlobals(tree.rootNode);
    expect(globals).toMatchSnapshot();
})