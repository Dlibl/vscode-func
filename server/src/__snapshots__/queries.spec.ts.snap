// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should query globals 1`] = `
Array [
  Object {
    "text": "id",
    "type": "globalVar",
  },
  Object {
    "text": "test",
    "type": "function",
  },
]
`;
